bedrock/firefox/tests/test_views.py

```diff
@@ -3,18 +3,160 @@
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 import json
+from urllib import unquote_plus
+from urlparse import parse_qs
 
 from django.test import override_settings
 from django.test.client import RequestFactory
 
 from bedrock.base.urlresolvers import reverse
-from mock import patch
+from mock import patch, Mock
 from nose.tools import eq_, ok_
 
 from bedrock.firefox import views
 from bedrock.mozorg.tests import TestCase
 
 
+@override_settings(STUB_ATTRIBUTION_HMAC_KEY='achievers')
+@patch.object(views, 'time', Mock(return_value=12345.678))
+class TestStubAttributionCode(TestCase):
+    def _get_request(self, params):
+        rf = RequestFactory()
+        return rf.get('/', params,
+                      HTTP_X_REQUESTED_WITH='XMLHttpRequest',
+                      HTTP_ACCEPT='application/json')
+
+    def test_not_ajax_request(self):
+        req = RequestFactory().get('/', {'source': 'malibu'})
+        resp = views.stub_attribution_code(req)
+        self.assertEqual(resp.status_code, 400)
+        data = json.loads(resp.content)
+        self.assertEqual(data['error'], 'Resource only available via XHR')
+
+    def test_no_valid_param_names(self):
+        req = self._get_request({'dude': 'abides'})
+        resp = views.stub_attribution_code(req)
+        self.assertEqual(resp.status_code, 400)
+        data = json.loads(resp.content)
+        self.assertEqual(data['error'], 'no params')
+
+    def test_no_valid_param_data(self):
+        params = {'utm_source': 'br@ndt', 'utm_medium': 'ae#her'}
+        req = self._get_request(params)
+        resp = views.stub_attribution_code(req)
+        self.assertEqual(resp.status_code, 400)
+        data = json.loads(resp.content)
+        self.assertEqual(data['error'], 'no params')
+
+    def test_some_valid_param_data(self):
+        params = {'utm_source': 'brandt', 'utm_content': 'ae#her'}
```

Comment:
Ah yes, good call. A better test would be a more likely invalid character. I'll fix.